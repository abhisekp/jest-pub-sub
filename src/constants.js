export const PUB_CHANNEL = 'test-channel';
