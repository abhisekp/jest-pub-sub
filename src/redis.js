import redis from 'redis';
import bluebird from 'bluebird';

bluebird.promisifyAll(redis.RedisClient.prototype);
bluebird.promisifyAll(redis.Multi.prototype);

export default redis;
