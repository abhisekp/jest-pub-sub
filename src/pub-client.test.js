import _ from 'lodash/fp';
import debug from 'debug';
import redis from './redis';
import { PUB_CHANNEL } from './constants';

const d = debug('app:pub-client');

const pubClient = redis.createClient();

let subClientAll;

beforeAll(() => {
  /* subscribe the channel */
  subClientAll = redis.createClient();
  subClientAll.subscribe(PUB_CHANNEL);
});

afterAll(() => {
  /* unsubscribe the channel */
  if (_.isObject(subClientAll) && _.isFunction(subClientAll.unsubscribe)) {
    subClientAll.unsubscribe(PUB_CHANNEL);
  }
});

describe('Publish Redis Channel', () => {
  test('Publish OK with no subscribers', async () => {
    const reply = await pubClient.publishAsync(PUB_CHANNEL, 'OK');
    d('Reply: %o', reply);
    expect(reply).toBe(0);
  });

  test('Publish OK with one subscriber', (done) => {
    const subClient = redis.createClient();

    /* subscribe the channel */
    subClient.subscribe(PUB_CHANNEL);

    /* listen to the subscribe event and publish OK */
    subClient.on('subscribe', async (channel, count) => {
      d('Channel %o, Count %o', channel, count);

      const reply = await pubClient.publishAsync(PUB_CHANNEL, 'OK');
      d('Reply: %o', reply);
      expect(reply).toBe(1);
    });

    /* listen to the published message */
    subClient.on('message', (channel, message) => {
      if (channel !== PUB_CHANNEL) return;

      expect(message).toMatchSnapshot();
      subClient.unsubscribe(PUB_CHANNEL);
      done();
    });
  });
});
