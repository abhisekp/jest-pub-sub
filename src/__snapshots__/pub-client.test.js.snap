// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publish Redis Channel Publish OK with one subscriber 1`] = `"OK"`;
